### PR TITLE
feat: auth handlers

### PR DESCRIPTION
**Pull request by solaris007**
Prepare authentication to support handlers. Handler support is needed for upcoming IMS auth while still supporting legacy key-based auth.

 Via such handlers we will be able to support a chain of fallback authentication methods. Handlers will set an auth info object which can be expanded in the future to feature roles and permissions.

### CODE REVIEW

File: src/support/auth/authentication-manager.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  /**
+   * Authenticate the request with all the handlers.
+   * @param {Object} request - The request object
+   * @param {UniversalContext} context - The context object
+   * @return {Promise<AuthInfo>} The authentication info
+   * @throws {NotAuthenticatedError} If no handler was able to authenticate the request
+   */
+  async authenticate(request, context) {
+    for (const handler of this.handlers) {
```
Comment (ekremney) [inline src/support/auth/authentication-manager.js:45]: looping through handlers doesn't seem to be the best way to all these. Especially when `handler.checkAuth` needs to perform some http call for an auth service like IMS.

wouldn't it be possible to detect what authenticating mechanism user wants to use? Ie. we can look at the `x-api-key` header. If it exists, then user wants to authenticate via legacy.

helix-admin does a similar thing: https://github.com/adobe/helix-admin/blob/main/src/support/auth.js#L560

File: src/index.js [hunk 2]
```diff
@@ -135,7 +137,7 @@ const { WORKSPACE_EXTERNAL } = SLACK_TARGETS;
 
 export const main = wrap(run)
   .with(dataAccess)
-  .with(auth)
+  .with(auth, { authHandlers: [LegacyApiKeyHandler, AdobeImsHandler] })
```
Comment (ekremney) [inline src/index.js:140]: providing a list of auth handler here seems a bit redundant. Do we have any case where we have a handler implemented, but not provided here?

maybe we can have controller specific handlers instead? Controllers can define which handlers they use?

Comment (solaris007) [reply src/index.js:140]: yes, we can do that, but not in step one please.

**Review by ekremney — APPROVED** [review]

File: src/index.js (hunk 2 above)
Comment (ekremney) [reply src/index.js:140]: will be expecting that in the step two then 🧖‍♂️

**Pull request merged**